pisek/visualize.py

```diff
@@ -13,304 +9,330 @@
+from math import ceil, inf
 import os
-import re
-import sys
-from colorama import Fore
-from typing import Optional, Union, Iterable, Callable
-
-from pisek.env.task_config import load_config, TaskConfig, SubtaskConfig
-from pisek.jobs.parts.task_job import TaskHelper
-
-VERDICTS = {
-    "ok": "·",
-    "partial": "P",
-    "timeout": "T",
-    "timeout_limited": "t",
-    "wrong_answer": "W",
-    "error": "!",
-}
-VERDICTS_ORDER = ["·", "P", "t", "T", "W", "!"]
-
-# subtask section
-TestCaseResult = namedtuple("TestCaseResult", ("name", "verdict", "value", "points"))
-
-
-def red(msg: str) -> str:
-    return f"{Fore.RED}{msg}{Fore.RESET}"
-
-
-def group_by_subtask(
-    results: list[TestCaseResult], config: TaskConfig
-) -> dict[int, list[TestCaseResult]]:
-    subtasks: dict[int, list[TestCaseResult]] = {
-        num: [] for num, _ in config.subtasks.items()
-    }
-    for result in results:
-        for i, subtask in config.subtasks.items():
-            if in_subtask(result.name, subtask):
-                subtasks[i].append(result)
-    return subtasks
-
+from typing import Optional
+
+from pisek.utils.text import pad, tab, colored
+from pisek.utils.terminal import terminal_width
+from pisek.jobs.parts.solution_result import Verdict
+from pisek.env.task_config import load_config, TaskConfig
+
+test_name_max_length = 0
+
+
+class MissingSolution(Exception):
+    def __init__(self, name: str) -> None:
+        super().__init__(f"Missing solution {name} in testing log.")
+
+
+@dataclass
+class LoggedResult:
+    verdict: Verdict
+    points: float
+    time: float
+    test: str
+    original_verdict: Verdict
+
+    def __init__(
+        self,
+        verdict: Verdict,
+        points: float,
+        measured_value: float,
+        test: str,
+        original_verdict: Verdict,
+    ) -> None:
+        global test_name_max_length
+
+        self.verdict = verdict
+        self.points = points
+        self.time = measured_value
+        self.test = test
+        self.original_verdict = original_verdict
+
+        test_name_max_length = max(test_name_max_length, len(self.test))
+
+    def to_str(
+        self,
+        limit: float,
+        segments: int,
+    ) -> str:
+        def get_bar() -> str:
+            percentage = self.time / limit
+            full = max(ceil(segments * percentage), 1)
+
+            bounded = min(full, segments)
+            overflown = max(0, full - segments)
+
+            overflown_max_length = terminal_width - segments - test_name_max_length - 40
+            if cut := (overflown > overflown_max_length):
+                overflown = overflown_max_length
+
+            if abs(percentage - 1) < 0.1:
+                color = "red"
+            elif abs(percentage - 1) < 0.5:
+                color = "yellow"
+            else:
+                color = "green"
+
+            full_bar = colored("━", color)
+
+            return (
+                f"[{full_bar*bounded}{'━'*(segments-bounded)}|"
```

Comment:
Idea: The non-colored bars could be a different character, say '┅' instead of '━'.

Reply:
I would keep `'┅'`'s special meaning to cutted bar.

Reply:
Whoops, I didn't notice that.